### ISSUE TITLE
Client Error when --max_size exceeds drive capacity 

### ISSUE THREAD

**Issue by pthalodezin** [Storj/dataserv-client#143]
using the farm command with v2.1.4
INFO dataserv_client.api 175: Starting build
INFO dataserv_client.builder 108: Resuming from height 116978
ERROR dataserv_client.builder 75: Failed to write shard, will try once more! 'IOError(28, 'No space left on device')'
Traceback (most recent call last):
  File "Scripts\dataserv-client", line 10, in <module>
    cli.main(sys.argv[1:])
  File "c:\Python27\lib\site-packages\dataserv_client\cli.py", line 190, in main
    return getattr(client, command_name)(**arguments)
  File "c:\Python27\lib\site-packages\dataserv_client\api.py", line 226, in farm
    set_height_interval=set_height_interval)
  File "c:\Python27\lib\site-packages\dataserv_client\api.py", line 198, in build
    rebuild=rebuild)
  File "c:\Python27\lib\site-packages\dataserv_client\builder.py", line 128, in build
    file_hash = self.generate_shard(seed, store_path, cleanup=cleanup)
  File "c:\Python27\lib\site-packages\dataserv_client\builder.py", line 77, in generate_shard
    RandomIO.RandomIO(seed).genfile(self.shard_size, path)
  File "c:\Python27\lib\site-packages\RandomIO\RandomIO.py", line 211, in genfile
    self.dump(f, size)
  File "c:\Python27\lib\site-packages\RandomIO\RandomIO.py", line 187, in dump
    fp.write(self.read(bufsz))
IOError: [Errno 0] Error

this occurs when the --max_size argument exceeds the capacity of the drive.  If the --max_size does not exceed the drive size, the farm command state that the build is finished and starts polling.  It would be ideal that the farm command goes to poll in either instance as the build is over at that point, either way.

**Comment by super3** [comment]
Please post the command you used.

**Comment by pthalodezin** [comment]
python Scripts\dataserv-client --config_path C:\Users\pthalodezin\.storj\config7.json --store_path Z:\FARM7 --max_size 7.132T --url=http://switch.driveshare.org farm  2>> W:\4_StorjLogs\FARM7\BUILD\storjlog7_B.log  

this is the successful command after I had multiple failed attempts at calculating the number.  This was the first number that worked as i decreased the value

**Comment by littleskunk** [comment]
In my Opionion it works as designed. A try catch would start polling but you would not know why it stoped building in the night. I dont think we should catch it.

**Comment by pthalodezin** [comment]
it's annoying, but it's nothing that was very hard to figure out.  trial and error at times.

**Comment by johnavgeros** [comment]
Can the client detect maximum capacity and auto set it correcting the user's entry if the user set larger than the maximum available?

**Comment by F483** [comment]
No it cannot detect if the given max size is larger then saved shards + free space. This is a bug and will be fixed.

**Comment by johnavgeros** [comment]
Thanks for clearing this out Fabian.

**Comment by super3** [comment]
@littleskunk I think this should be something you can handle. Check if the user has enough space to handle the build before starting. Return an error if they do not.

**Comment by littleskunk** [comment]
@pthalodezin No more free space is only one reason for IOError. If i understand the bug right you would like to see it polling no matter what exception build was throwing.

For the free space i would like to write a new issue as enhancement. The user should have the option to set the free space he would like to have. Build has to stop at that limit.

**Comment by pthalodezin** [comment]
@littleskunk I understood that there were many errors that could stop a build, but it is this specific error that is of concern.  I don't think that a blanket - keep polling exception would be beneficial, at least without logging and other notifications.  

I think the freespace enhancement would serve some but not all that have these errors.  The pre-check idea is more in line.  Check shard space + freespace and return error prior to building (preferably with a number that could be used).  

My bug is more focused on those wanting to fill drives to max capacity.  We all know that drive manufacturers advertise the number that looks the best (4TB drive does not hold 4TB) but there is also the operating system reserved space for maintenance (especially on NAS devices) that drives that number still lower.  A check that would at minimum throw an error that says "You do not have enough free drive space for that maximum size.  Reduce your max size and try again" would be beneficial.  That would save from the frustration of building and then having the client error out at the end of the build.

**Comment by nrenich** [comment]
I'm not sure of the exact approach programmatically, but I have always thrown a higher then normal max cap so that I get every possible byte out of my builds. I know it throws an error, even when I restart the batch file, but then I start polling and it's right as rain. I guess something more graceful could be in order, I'm just giving feedback. But don't throw out the idea that larger farmers will want an easy way to get max cap on a destination without manually determining it through trial and error.

**Comment by pthalodezin** [comment]
@nrenich That is exactly my thinking as well, but the issue lies in how you are running the different commands.  My understanding is that in the future, the use of the farm command will be necessary to ensure that audits work (audit being a function of farm).  If you are running poll (being that poll alone only pings the server that you are live and doesn't report anything on height), you won't be doing audits and essentially that would be a failed audit and the server could see your build as invalid.  The fix has to allow farm to start polling even if it hasn't reached the shard that it knows to be the last in the stack to reach --max_size.  

Since it appears that the build portion of farm builds until it reaches a predetermined hash/shard, I think it is right that the warning needs to come prior to actually building, so that the farmer can set the right max-size, telling the program when to stop.  

The rub is when someone is using a drive that has variable free capacity and that number changes from the beginning to the end of the build.  What isn't being talked about so far is that the ideal farmer would have a fixed capacity.  To have a variable capacity, especially going from larger to smaller is not going to be appealing to users of a live system.

**Comment by littleskunk** [comment]
@pthalodezin 
Lets try it with some simple examples.

User 1 has 100GB free space and starts building with --max_size=100G. 800 shards are possible. User 1 download some files, software update, videos what ever. At the end he will get IOError even if he had the needed space at the beginning. I would expect the exception. No need to change something.

User 2 has 100GB free space and a ntfs drive. User 3 has 100GB free space and a FAT32 drive. They will end up with different height. That is a porblem for the calculation.

User 4 would like to write as many shards as possible. Lets give him the possibility to define the space he would like to keep free. Build will check for every shard if there is enough space left.

I dont think a pre check of the free space will change anything. I only see that one possibility.

**Comment by littleskunk** [comment]
fixed in the next dataserv-client version (current version: 2.1.4). Unit test and documentation fixed as well.

You can set the minimum free space with --min_free_size=10G. Default is 1G. 0 is not allowed. The free space calculation has a short delay. It will stop a few file to late. With 8 workers i gues 1G is to low. As default value for 1 worker it should work.

**Comment by littleskunk** [comment]
Fixed the worker problem as well. Build will now stop as soon as free space + (active workers * shard size) hits the given free space limit. With more workers it will stop a little bit to early because the calculation counts the number of active threads * 128MB even if all workers only need 1 MB to finish there work.
